From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nathan Sanchez <nathanasanchez@google.com>
Date: Thu, 15 Oct 2026 11:24:36 -0400
Subject: [M156] [Vulkan] Treat VK_ERROR_UNKNOWN as VK_ERROR_DEVICE_LOST

Original change's description:
> [Vulkan] Treat VK_ERROR_UNKNOWN as VK_ERROR_DEVICE_LOST
>
> We normally would not expect VK_ERROR_UNKNOWN as this means there is
> either a problem with our calls to Vulkan which wasn't caught with
> validation layers, or the driver has an implementation issue. Since it
> is not normally expected and is unclear what the issue is we will treat
> is as a device lost.
>
> Bug: 568936430
> Change-Id: I87a145a418fd41a2e6d513ca1a7525556a55c3db
> Reviewed-on: https://skia-review.googlesource.com/c/skia/+/1395157
> Commit-Queue: Nathan Sanchez <nathanasanchez@google.com>
> Reviewed-by: Michael Ludwig <michaelludwig@google.com>

(cherry picked from commit ca5696a110e894f3a2cc0838b106cdd09eb9a965)

Bug: 575704689,568936430
Change-Id: I87a145a418fd41a2e6d513ca1a7525556a55c3db
Reviewed-on: https://skia-review.googlesource.com/c/skia/+/1400576
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: Chrome Cherry Picker <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>

diff --git a/src/gpu/ganesh/vk/GrVkGpu.cpp b/src/gpu/ganesh/vk/GrVkGpu.cpp
index e41868a1ef484bf4706807047973beeb6c36d0e2..0c7e4193c79f8486a262e17fe894bee3fbe58922 100644
--- a/src/gpu/ganesh/vk/GrVkGpu.cpp
+++ b/src/gpu/ganesh/vk/GrVkGpu.cpp
@@ -2658,6 +2658,7 @@ bool GrVkGpu::checkVkResult(VkResult result) {
         case VK_SUCCESS:
             return true;
         case VK_ERROR_DEVICE_LOST:
+        case VK_ERROR_UNKNOWN:
             if (!fDeviceIsLost) {
                 // Callback should only be invoked once, and device should be marked as lost first.
                 fDeviceIsLost = true;
@@ -2665,7 +2666,8 @@ bool GrVkGpu::checkVkResult(VkResult result) {
                                                 device(),
                                                 fDeviceLostContext,
                                                 fDeviceLostProc,
-                                                vkCaps().supportsDeviceFaultInfo());
+                                                vkCaps().supportsDeviceFaultInfo(),
+                                                /*isUnknownError=*/result == VK_ERROR_UNKNOWN);
             }
             return false;
         case VK_ERROR_OUT_OF_DEVICE_MEMORY:
diff --git a/src/gpu/graphite/vk/VulkanSharedContext.cpp b/src/gpu/graphite/vk/VulkanSharedContext.cpp
index 2b27b49674d2b8ceda436720d165dcdf59faa100..4ddcce660e58411e491aa25beef1f770c43e2878 100644
--- a/src/gpu/graphite/vk/VulkanSharedContext.cpp
+++ b/src/gpu/graphite/vk/VulkanSharedContext.cpp
@@ -268,6 +268,7 @@ bool VulkanSharedContext::checkVkResult(VkResult result) const {
     case VK_SUCCESS:
         return true;
     case VK_ERROR_DEVICE_LOST:
+    case VK_ERROR_UNKNOWN:
         {
             SkAutoMutexExclusive lock(fDeviceIsLostMutex);
             if (fDeviceIsLost) {
@@ -282,7 +283,8 @@ bool VulkanSharedContext::checkVkResult(VkResult result) const {
                                         device(),
                                         fDeviceLostContext,
                                         fDeviceLostProc,
-                                        vulkanCaps().supportsDeviceFaultInfo());
+                                        vulkanCaps().supportsDeviceFaultInfo(),
+                                        /*isUnknownError=*/result == VK_ERROR_UNKNOWN);
         return false;
     case VK_ERROR_OUT_OF_DEVICE_MEMORY:
     case VK_ERROR_OUT_OF_HOST_MEMORY:
diff --git a/src/gpu/vk/VulkanUtilsPriv.cpp b/src/gpu/vk/VulkanUtilsPriv.cpp
index 0c17025b82666a9a913b87185caa9bf72cf8ed6b..943f615eec4efd2e8868b824c28600cfb4825096 100644
--- a/src/gpu/vk/VulkanUtilsPriv.cpp
+++ b/src/gpu/vk/VulkanUtilsPriv.cpp
@@ -232,7 +232,8 @@ void InvokeDeviceLostCallback(const skgpu::VulkanInterface* vulkanInterface,
                               VkDevice vkDevice,
                               skgpu::VulkanDeviceLostContext deviceLostContext,
                               skgpu::VulkanDeviceLostProc deviceLostProc,
-                              bool supportsDeviceFaultInfoExtension) {
+                              bool supportsDeviceFaultInfoExtension,
+                              bool isUnknownError) {
     if (!deviceLostProc) {
         return;
     }
@@ -241,6 +242,15 @@ void InvokeDeviceLostCallback(const skgpu::VulkanInterface* vulkanInterface,
     std::vector<VkDeviceFaultVendorInfoEXT> vendorInfos = {};
     std::vector<std::byte> vendorBinaryData = {};
 
+    if (isUnknownError) {
+        deviceLostProc(deviceLostContext,
+                       "No details: A VK_UNKNOWN_ERROR has occured.",
+                       addressInfos,
+                       vendorInfos,
+                       vendorBinaryData);
+        return;
+    }
+
     if (!supportsDeviceFaultInfoExtension) {
         deviceLostProc(deviceLostContext,
                        "No details: VK_EXT_device_fault not available/enabled.",
diff --git a/src/gpu/vk/VulkanUtilsPriv.h b/src/gpu/vk/VulkanUtilsPriv.h
index c9cadcdf4c33de606e32ab12c41522d747dce93e..9865b0d7c21e3f0c35a0f85c71eac3f059376a1c 100644
--- a/src/gpu/vk/VulkanUtilsPriv.h
+++ b/src/gpu/vk/VulkanUtilsPriv.h
@@ -39,9 +39,9 @@ struct ShaderCaps;
 namespace skgpu {
 
 class ShaderErrorHandler;
-struct VulkanInterface;
-struct VulkanBackendContext;
 class VulkanExtensions;
+struct VulkanBackendContext;
+struct VulkanInterface;
 
 enum VkVendor {
     kAMD_VkVendor = 0x1002,
@@ -362,15 +362,17 @@ bool AllocateAndBindImageMemory(skgpu::VulkanAlloc* outVulkanAlloc,
 #endif // SK_BUILD_FOR_ANDROID
 
 /**
- * Calls faultProc with faultContext; passes debug info if VK_EXT_device_fault is supported/enabled.
+ * Calls faultProc with faultContext; passes debug info if VK_EXT_device_fault is supported/enabled
+ * and the error is VK_ERROR_DEVICE_LOST.
  *
- * Note: must only be called *after* receiving VK_ERROR_DEVICE_LOST.
+ * Note: must only be called *after* receiving VK_ERROR_DEVICE_LOST or VK_ERROR_UNKNOWN.
  */
 void InvokeDeviceLostCallback(const skgpu::VulkanInterface* vulkanInterface,
                               VkDevice vkDevice,
                               skgpu::VulkanDeviceLostContext faultContext,
                               skgpu::VulkanDeviceLostProc faultProc,
-                              bool supportsDeviceFaultInfoExtension);
+                              bool supportsDeviceFaultInfoExtension,
+                              bool isUnknownError);
 
 sk_sp<skgpu::VulkanInterface> MakeInterface(const skgpu::VulkanBackendContext&,
                                             const skgpu::VulkanExtensions* extOverride,
